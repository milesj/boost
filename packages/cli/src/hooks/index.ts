export * from './useDimensions';
export * from './useIsMounted';
export * from './useListNavigation';
export * from './useProgram';
export * from './useRenderLoop';
