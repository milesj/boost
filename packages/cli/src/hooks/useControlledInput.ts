import { useInput } from 'ink';

export interface ControlledInputOptions {
  focused?: boolean;
  onBackspace?: () => void;
  onDelete?: () => void;
  onEscape?: () => void;
  onInput?: (value: string) => void;
  onKeyDown?: () => void;
  onKeyLeft?: () => void;
  onKeyRight?: () => void;
  onKeyUp?: () => void;
  onPageDown?: () => void;
  onPageUp?: () => void;
  onReturn?: () => void;
  onTab?: () => void;
}

export default function useControlledInput({
  focused,
  onBackspace,
  onDelete,
  onEscape,
  onInput,
  onKeyDown,
  onKeyLeft,
  onKeyRight,
  onKeyUp,
  onPageDown,
  onPageUp,
  onReturn,
  onTab,
}: ControlledInputOptions) {
  useInput(
    // eslint-disable-next-line complexity
    (input, key) => {
      if (key.escape || (key.ctrl && input === 'c')) {
        onEscape?.();

        return;
      }

      if (key.upArrow) {
        onKeyUp?.();
      } else if (key.downArrow) {
        onKeyDown?.();
      } else if (key.leftArrow) {
        onKeyLeft?.();
      } else if (key.rightArrow) {
        onKeyRight?.();
      } else if (key.pageUp) {
        onPageUp?.();
      } else if (key.pageDown) {
        onPageDown?.();
      } else if (key.return) {
        onReturn?.();
      } else if (key.tab) {
        onTab?.();
      } else if (key.backspace) {
        onBackspace?.();
      } else if (key.delete) {
        onDelete?.();
      } else {
        onInput?.(input);
      }
    },
    { isActive: focused },
  );
}
