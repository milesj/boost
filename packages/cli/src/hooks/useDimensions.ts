import { useEffect, useState } from 'react';
import { useStdout } from 'ink';
import { screen } from '@boost/terminal';

export function useDimensions(): { height: number; width: number } {
	const [dims, setDims] = useState(() => screen.size());
	const { stdout } = useStdout();

	useEffect(() => {
		// istanbul ignore next
		const handler = () => {
			setDims(screen.size());
		};

		stdout?.on('resize', handler);

		return () => {
			stdout?.off('resize', handler);
		};
	}, [stdout]);

	return {
		height: dims.rows,
		width: dims.columns,
	};
}
