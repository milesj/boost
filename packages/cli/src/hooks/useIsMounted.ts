import { useEffect, useRef } from 'react';

export function useIsMounted() {
	const mounted = useRef(false);

	useEffect(() => {
		mounted.current = true;

		return () => {
			mounted.current = false;
		};
	}, []);

	return mounted;
}
