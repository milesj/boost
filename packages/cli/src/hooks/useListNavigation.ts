import { useCallback, useState } from 'react';

export function useListNavigation(
	list: { disabled?: boolean; divider?: boolean }[],
	defaultIndex?: number,
) {
	const getNextIndex = useCallback(
		(index: number, step: number) => {
			let nextIndex = index;

			if (nextIndex >= list.length) {
				nextIndex = 0;
			} else if (nextIndex < 0) {
				nextIndex = list.length - 1;
			}

			if (list[nextIndex]?.disabled ?? list[nextIndex]?.divider) {
				nextIndex = getNextIndex(nextIndex + step, step);
			}

			return nextIndex;
		},
		[list],
	);

	const [highlightedIndex, setHighlightedIndex] = useState(
		() => defaultIndex ?? getNextIndex(0, 1),
	);

	const onKeyDown = useCallback(() => {
		setHighlightedIndex((index) => getNextIndex(index + 1, 1));
	}, [getNextIndex]);

	const onKeyUp = useCallback(() => {
		setHighlightedIndex((index) => getNextIndex(index - 1, -1));
	}, [getNextIndex]);

	const onKeyLeft = useCallback(() => {
		setHighlightedIndex(getNextIndex(0, 1));
	}, [getNextIndex]);

	const onKeyRight = useCallback(() => {
		setHighlightedIndex(getNextIndex(list.length - 1, -1));
	}, [getNextIndex, list]);

	return {
		highlightedIndex,
		onKeyDown,
		onKeyLeft,
		onKeyRight,
		onKeyUp,
	};
}
