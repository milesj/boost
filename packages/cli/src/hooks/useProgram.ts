import { useCallback, useContext } from 'react';
import { useApp } from 'ink';
import { ExitError } from '@boost/common';
import { EXIT_FAIL, EXIT_PASS } from '../constants';
import { ProgramContext } from '../ProgramContext';
import { ProgramContextType } from '../types';

export function useProgram(): ProgramContextType {
	const { exit: appExit } = useApp();
	const program = useContext(ProgramContext);

	const exit = useCallback(
		(error?: Error | string, code?: number) => {
			if (error) {
				appExit(new ExitError(error instanceof Error ? error.message : error, code ?? EXIT_FAIL));
			} else {
				appExit(new ExitError('', EXIT_PASS));
			}
		},
		[appExit],
	);

	return { ...program, exit };
}
