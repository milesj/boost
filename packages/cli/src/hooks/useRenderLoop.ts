import { useCallback, useEffect, useReducer, useRef } from 'react';

// eslint-disable-next-line no-magic-numbers
const DEFAULT_FPS = 1000 / 30;

export function useRenderLoop(fps: number = DEFAULT_FPS): () => void {
	const [, forceUpdate] = useReducer((count) => count + 1, 0);
	const timer = useRef<NodeJS.Timeout>();

	const clear = useCallback(() => {
		if (timer.current) {
			clearInterval(timer.current);
		}
	}, []);

	useEffect(() => {
		timer.current = setInterval(forceUpdate, fps);

		return clear;
	}, [clear, fps]);

	return clear;
}
