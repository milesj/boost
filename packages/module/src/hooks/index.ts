// https://nodejs.org/api/module.html#customization-hooks

import { register } from 'node:module';

register('./hook-typescript.mjs', import.meta.url);
