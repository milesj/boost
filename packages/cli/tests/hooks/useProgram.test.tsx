import React, { useEffect } from 'react';
import { Box } from 'ink';
import { render } from 'ink-testing-library';
import { useProgram } from '../../src/hooks/useProgram';
import { describe, it, expect } from 'vitest';

describe('Input', () => {
	function TestProgram() {
		const program = useProgram();

		useEffect(() => {}, [program.exit]);

		return <Box />;
	}

	it('custom `exit` doesnt cause an infinite loop', () => {
		const { lastFrame } = render(<TestProgram />);

		expect(lastFrame()).toMatchSnapshot();
	});
});
