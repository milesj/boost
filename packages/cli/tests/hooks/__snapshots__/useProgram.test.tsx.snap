// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input custom \`exit\` doesnt cause an infinite loop 1`] = `""`;
